packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
+ * } NestedData
+ * Corresponding PassStyle values are
+ * the PassStyles of LeafData, "copyArray", "copyRecord", "copyError"
+ *
+ * NestedData is parameterized over various "deep" constraints. See
+ * the discussion in CopyArray and CopyRecord above.
  */
 
 /**
- * @typedef {Promise | Remotable} PassableCap
- * The leaves of a Passable's pass-by-copy superstructure.
+ * @typedef {NestedData<PureData>} PureData
+ * Contains no capabilities (promises, remotables)
+ * Contains no sets or maps
+ * Contains no proxies, or anything which can sense being examined.
+ * Cannot be tested for, but only assured by construction
+ * pureCopy(OnlyData) => PureData
+ */
+
+/**
+ * @typedef {PureData | NestedData<OnlyData>} OnlyData
+ * Tested by `isOnlyData`, `assertOnlyData`
+ * Contains no capabilities (promises, remotables)
+ * Contains no sets or maps
+ * The objects it contains may contain proxies, and therefore
+ * may cause side effects when examined.
+ */
+
+/**
+ * @typedef {Remotable | OnlyData | NestedData<PassableKey> } PassableKey
+ *
+ * PassableKeys can be compared for equivalence with `sameKey`. A
+ * PassableKey can be used as a key in a StrongStore, i.e., used
+ * as a key in a PassableMap, MemoryMap, or VirtualMap, and used as
+ * an element of a PassableSet, MemorySet, or VirtualSet.
+ *
+ * Note that a PassableKey cannot contain promises nor passable
+ * stores.
+ *
+ * Passing should preserve `sameKey` equivalence. Given that PassableKeys
+ * `Xa` and `Ya` in vat A, when passed to vat B, arrive as `Xb` and `Yb`, then
+ * `sameKey(Xa, Ya)` in vat A must have the same answer as `sameKey(Xb, Yb)`
+ * in vat B.
+ */
+
+/**
+ * @template {PassableKey} K
+ * @typedef {*} PassableStore
+ * @property {(key: K) => boolean} has
+ * @property {() => K[]} keys
+ *
+ * A PassableStore may contain PassableKeys that cannot be sorted, so
+ * the distributed object semantics itself does not specify a deterministic
+ * enumeration order for passable stores. However, a given programming
+ * language should specify a deterministic enumeration order. For JS
+ * this would likely be based on insertion order.
+ *
+ * PassableStores are not themselves
+ * PassableKeys and so cannot be compared with `sameKey` nor serve as
+ * indexes in other passable stores. However, they are Comparable and
+ * so may still be compared for equivalence with `sameStructure`. Two
+ * PassableStores that are equivalent by `sameStructure` may have
+ * different enumeration orders according to the more specific language
+ * binding semantics.
+ */
+
+/**
+ * @template {PassableKey} K
+ * @typedef {PassableStore<K>} PassableSet
+ */
+
+/**
+ * @template {PassableKey} K
+ * @template {Passable} V
+ * @typedef {PassableStore<K>} PassableMap
+ * @property {(key: K) => V} get
+ * @property {() => V[]} values
+ * @property {() => [K,V][]} entries
+ */
+
+/**
+ * @template T
+ * @typedef { PassableKey | PassableSet<T> | PassableMap<T,T> } NestedComparable
```

Comment:
I don't think you really mean a passable map of exactly T mapped to T.   These are not nominal types, so they must be specialisable:

```suggestion
 * @template K, V
 * @typedef { PassableKey | PassableSet<K> | PassableMap<K,V> } NestedComparable
```